File: training-data/erlang/bubble_sort.erl
-module(bubble_sort).
-export([sort/1]).

sort(List) ->
    sort(List, length(List)).

sort(List, 0) -> List;
sort(List, N) ->
    {NewList, _} = pass(List),
    sort(NewList, N - 1).

pass([A, B | Rest]) when A > B ->
    {Sorted, Changed} = pass([A | Rest]),
    {[B | Sorted], Changed};
pass([A, B | Rest]) ->
    {Sorted, Changed} = pass([B | Rest]),
    {[A | Sorted], Changed};
pass(List) ->
    {List, false}.

main() ->
    io:format("~p~n", [sort([5, 2, 9, 1, 5, 6])]).

File: training-data/haskell/BubbleSort.hs
bubbleSort :: [Int] -> [Int]
bubbleSort xs = if sorted == xs then xs else bubbleSort sorted
  where
    sorted = onePass xs
    onePass (a:b:rest)
      | a > b = b : onePass (a : rest)
      | otherwise = a : onePass (b : rest)
    onePass xs' = xs'

main :: IO ()
main = print (bubbleSort [5, 2, 9, 1, 5, 6])
